File: solutions/0049-group-anagrams/solution.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> groups;
        for (const string& s : strs) {
            string key = s;
            sort(key.begin(), key.end());
            groups[key].push_back(s);
        }
        vector<vector<string>> res;
        for (auto& p : groups) res.push_back(p.second);
        return res;
    }
};

int main() {
    Solution sol;
    vector<string> in = {"eat", "tea", "tan", "ate", "nat", "bat"};
    for (auto& g : sol.groupAnagrams(in)) {
        for (auto& s : g) cout << s << " ";
        cout << endl;
    }
}

File: solutions/0049-group-anagrams/Solution.java
import java.util.*;

public class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> groups = new HashMap<>();
        for (String s : strs) {
            char[] chars = s.toCharArray();
            Arrays.sort(chars);
            groups.computeIfAbsent(new String(chars), k -> new ArrayList<>()).add(s);
        }
        return new ArrayList<>(groups.values());
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.groupAnagrams(
            new String[]{"eat", "tea", "tan", "ate", "nat", "bat"}));
    }
}
